### ISSUE TITLE
[DataCatalog2.0]: Update `KedroDataCatalog` CLI logic and make it reusable

### ISSUE THREAD

**Issue by noklam** [kedro-org/kedro#3312]
## Description
Parent issue: https://github.com/kedro-org/kedro/issues/4472

Suggested plan: https://github.com/kedro-org/kedro/issues/3312#issuecomment-2612417154

## Context

Background: https://linen-slack.kedro.org/t/16064885/when-i-say-catalog-list-in-a-kedro-jupter-lab-instance-it-do#ad3bb4aa-f6f9-44c6-bb84-b25163bfe85c

With dataset factory, the "defintion" of a dataset is not known until the pipeline is run. When user is using a Jupyter notebook, they expected to see the full list of dataset with `catalog.list`.

Current workaround to see the datasets for `__default__` pipeline look like this:

```python
for dataset in pipeline["__default__"].data_sets():
  catalog.exists(dataset)
```

When using the CLI commands, e.g. `kedro catalog list` we do matching to figure out which factory mentions in the catalog match the datasets used in the pipeline, but when going through the interactive flow no such checking has been done yet. 

## Possible Implementation
Could check dataset existence when the session is created. We need to verify if that has any unexpected side effects.

This ticket is still open scope and we don't have a specify implementation in mind. The person who pick up can evaluate different approaches, with considerations of side-effect, avoid coupling with other components.

## Possible Alternatives

- `catalog.list( pipeline=<name>`) - not a good solution because catalog wouldn't have access to a pipeline
- Do something similar to what's happening when `kedro catalog list` is called.

**Comment by datajoely** [comment]
could we have something like `catalog.resolve(pipeline:Optional[str]).list()`?

**Comment by merelcht** [comment]
This Viz issue is related: https://github.com/kedro-org/kedro-viz/issues/1480

**Comment by MarcelBeining** [comment]
> could we have something like `catalog.resolve(pipeline:Optional[str]).list()`?

That would be perfect! We would need such a thing

**Comment by noklam** [comment]
@MarcelBeining Can you explains a bit more why you need this? I am thinking about this again because I am trying to build a plugin for kedro and this would come in handy to compile a static version of configuration.

**Comment by noklam** [comment]
@MarcelBeining Did I understand this question correctly as:
- Find which datasets is not written in `catalog.yml` yet? I have some WIP in https://github.com/noklam/kedro-inspect which explores this idea but I haven't finished it.
- 

Does `kedro catalog resolve` or `kedro catalog list` helps you? If not what are missing?

**Comment by MarcelBeining** [comment]
@noklam "Find which datasets is not written in catalog.yml including dataset factory resolves, yet" , yes

kedro catalog resolve shows what I need, but it is a CLI command and I need it within Python (of course one could use os.system etc, but a simple extension of catalog.list() should not be that hard)

**Comment by noklam** [comment]
@MarcelBeining Are you integrating this with some extra functionalities? How do you consume this information?

**Comment by ianwhale** [comment]
@noklam 

Adding on from our discussion on slack,

`kedro catalog resolve` does what I'd want. 

But I'd also like that information easily consumable in a notebook (for example).

So if my catalog stores models like:

```yaml
"{experiment}.model":
  type: pickle.PickleDataset
  filepath: data/06_models/{experiment}/model.pickle
  versioned: true
```

I would want to be able to (somehow) do something like:

```python
models = {}

for model_dataset in [d for d in catalog.list(*~*magic*~*) if ".model" in d]:
    models[model_dataset] = catalog.load(model_dataset)
```

Its a small thing. But I was kind of surprised to not see my `{experiment}.model` entries not listed at all in `catalog.list()`.

**Comment by noklam** [comment]
Another one, I drag it back to `inbox` (the original priority is `medium`)
https://kedro-org.slack.com/archives/C03RKP2LW64/p1716549750334469

**Comment by noklam** [comment]
What if `DataCatalog` is iterable?

```python
for datasets in data_catalog:
   ...
```

**Comment by datajoely** [comment]
I think it's neat @noklam , but I don't know if it's discoverable. 

To me `DataCatalog.list()` feels more powerful in the IDE than `list(DataCatalog)`...

**Comment by astrojuanlu** [comment]
why_not_both.gif

```
def list(self):
    ...

def __iter__(self):
    return self.list()
```

**Comment by Galileo-Galilei** [comment]
I've also wanted to be able to iterate through the datasets for a while, but it raises some unanswered questions: 
- should we iterate on ``catalog`` (maybe more intuitive) or ``catalog.datasets`` as described in https://github.com/kedro-org/kedro/issues/3916 (maybe more accurate, especially in regard to the "resolving" issue discussed below) ?
-  How does this loop would handle dataset factory? We could eventually replace : 
    - ``catalog.list()`` by ``[dataset.name for dataset in catalog]`` 
    - ``catalog.search`` (which does not exist but is suggested in #3917) by ``[dataset.name for dataset in catalog if dataset.name.match(regex)]``

But we always face the same issue: we would need to "resolve" the dataset factory first relatively to a pipeline. it would eventually give: ``[dataset.name for dataset in catalog.resolve(pipeline)]``, but is it really a better / more intuitive syntax ?

**Comment by noklam** [comment]
Catalog.list already support regex, isn't that identical to what you suggest as catalog.search?

**Comment by datajoely** [comment]
@noklam you can only search by name, namespaces aren't really supported and you can't search by attribute

**Comment by ElenaKhaustova** [comment]
Inconsistency between CLI commands and interactive workflow is a problem related to all `kedro catalog` commands. Currently, all CLI logic is implemented within commands and cannot be reused across the codebase. This is confusing for users as one cannot perform the same actions that are available via commands in the interactive workflow. It also requires additional maintenance as any catalog-related changes should be validated for CLI logic as well.

Thus, we suggest refactoring CLI logic and moving it to the `session` level so that it can be reused for CLI and interactive workflows. We also suggest doing that together with the deprecation of the old catalog since CLI logic currently relies on some private fields and methods of the old catalog that will be removed.

We also think that we should not couple catalog and pipelines, so we do not consider extending `catalog.list()` method for that and suggest using `session` instead.

**Comment by datajoely** [comment]
Which I realise this is exactly what I pitched 18 months ago 😂 
https://github.com/kedro-org/kedro/issues/3312#issuecomment-1812517010

**Comment by ElenaKhaustova** [comment]
@datajoely
I think the conceptual problem here is that we were not clear enough in explaining to users that datasets **are not** equal to patterns. After we added `CatalogConfigResolver` patterns can be listed like that for both new and old catalogs: `catalog.config_resolver.list_patterns()`.

The idea of listing datasets and patterns together looks interesting. But doing so could lead to confusion, as people may not differentiate datasets and patterns. So, providing an interface to access both separately and communicate this better seems more reasonable to me.

There is also a high chance that `catalog.list()` will be replaced with another filtering method or removed, as now we have a dict-like interface to list keys, values, and items: https://github.com/kedro-org/kedro/issues/3917. 

I see the pain point about the discrepancy between `catalog.list()` and the CI command `kedro catalog list`, where the last considers the pipeline but the first does not. But as suggested above we think this can be solved via `session`, not `catalog` to avoid coupling `catalog` and `pipelines`.

Long story short - we suggest:

1. Keep `catalog.config_resolver.list_patterns()` to list patterns
2. Keep dict-like interface to list datasets
3. Replace `catalog.list()` with `catalog.filter()`or remove it
4. Move CI logic to the session, so it is possible to list datasets with regards to pipelines from the CI and in the interactive workflow, something like that: `session.catalog_list()`
5. Communicate better difference between datasets and patterns, resolution based on pipelines

These 5 should address the points you mentioned as well as others we discussed above.

**Comment by datajoely** [comment]
Okay we're actually designing for different things. Perhaps we could print out some of these recommendations when the ipython extension is loaded because even experienced kedro users are going to know how to retrieve patterns etc

**Comment by datajoely** [comment]
Now this is prioritised for the 1.0.0 collection of breaking changes I want to push that the current CLI command is useless:

<img width="479" alt="Image" src="https://github.com/user-attachments/assets/46e9ad79-b7a1-4df9-b9b0-0121623734e9" />

I no longer have access to telemetry, but last time I looked at it was barely used.

What it does well:

- It interpolates the dataset factories so you don't get the situation where `catalog.list()` doesn't work until the dataset has been discovered eagerly.
- It breaks down the datasets by type, but this is now solved in the python API

What it doesn't do well:

- It's not really machine readable, the output is technically YAML, but it's hard to do anything with it as the logging cannot be disabled by the user from the CLI on the own
- I think a detailed JSON output would be amazing, being able to do `kedro catalog list | jq` would be a productivity driver for users.

**Comment by ElenaKhaustova** [comment]
@datajoely, yep, now it's a good time to point to possible improvements. Can you please elaborate on what you mean by `detailed`? Do you expect any other information included?

**Comment by datajoely** [comment]
I a JSON structure something like this would be more useful from a machine interpretability point of view:

```json
[
   {"dataset_name": "...", "dataset_type": "...", "pipelines":["...","..."] },
]
```

There is more that is possibly useful such as classpath for custom datasets, which YAML file it was found in etc.

**Comment by astrojuanlu** [comment]
Thanks for the long writeup @ElenaKhaustova ! Left an idea about the interactive functionality in https://github.com/kedro-org/kedro/issues/4481#issuecomment-2707044866. I understand it's a thorny issue, hope we can unblock it.

About your "current vs updated functionality", just to clarify: Is the idea to keep the `kedro catalog rank`, `kedro catalog resolve`, `kedro catalog list` CLI commands just as they are now, without user changes? If so, the rationale to move their logic to the session looks good to me, as well as the programmatic APIs.

**Comment by Galileo-Galilei** [comment]
Hi, thank you very much for explaining everything at stake here. I am not familiar yet with the suggested implementation yet and I am not sure what I'll suggest hereafter completly makes sense, so take it with a grain of salt.

### Catalog level vs session level command

> **Why not move catalog commands to catalog level**
> We could implement commands at the catalog level, but that would lead to coupling catalog and pipelines, which is not desired.

This is definitely what bother me the most in the suggested implementation. From a end user / plugin developer experience, I've almost never seen anyone using the session interactively forservearl reasons: 
- This is **painful to use interactively** because the session encapsulates a lot of thing and you lose finer control, and **I've seen a lot of user use ``pipeline`` + ``catalog`` + ``runner`` in interactive mode rather than using the session directly**. There are also a lot of **plugins** using ``dataset.exist()``  to materialize dataset factories in a catalog instance, and **I am not sure they will be able to do it easily without creating a session**. I think it would be much easier to interact with the standalone catalog directly. 
- We have an **history of users asking for a standalone use of catalog**, e.g. https://github.com/kedro-org/kedro/discussions/4273 or https://github.com/kedro-org/kedro/issues/3941. They may, or may not, create pipeline and want to persist catalog factories by handling them at a catalog level, albeit maybe in this case it would be better not to be polluted with these methods in case they don't use pipelines.  
- Using the session is a **pain for testing because you need a lot of monkeypatching**, and using catalog directly is *much* easier. 

➡ I don't say we *must* do it at the catalog level rather than the session level, just that it would let some of our important catalog-related issues unsolved, so we should really thought about it. 

 
### Commands review
 
The argument of "coupling catalog with pipelines" is valid for _some_ methods, but I don't think it is for *all* methods: 
 
1.  ``kedro catalog rank`` -> ``list_catalog_patterns`` definitely looks like a catalog level method, not a session level one. Your code example gives a good hint: 

```python
kedro ipython
In [1]: session.list_catalog_patterns()
Out[1]: ['{name}.{folder}#csv', '{name}_data', 'out-{dataset_name}', '{dataset_name}#csv', 'in-{dataset_name}']

In [2]: runtime_pattern = {"{default}": {"type": "MemoryDataset"}}

In [3]: catalog.config_resolver.add_runtime_patterns(runtime_pattern)

In [4]: session.list_catalog_patterns()
Out[4]: ['{name}.{folder}#csv', '{name}_data', 'out-{dataset_name}', '{dataset_name}#csv', 'in-{dataset_name}', '{default}']
```

The runtime pattern is added by a method of an attribute of the catalog, but we would not be able to list them at the catalog level? This seems very weird: ``catalog.list_patterns()`` looks much more intuitive. 

2. ``kedro catalog resolve`` -> `` def resolve_catalog_patterns(self, include_default: bool = False) -> dict[str, Any]: ``

I am not sure of what we want this method to do: 
- if we want it to modify the catalog inplace to resolve all factories (as you currently do), I am in favor of putting it at the session level 
- if we want it to return a copy of the current catalog with resolved factories, what is the problem of letting it at the catalog level? As someone using the catalog as a standalone component (e.g. in a plugin) I can imagine a bunch os situations where I want to do  :  ``resolved_catalog=catalog.resolve(pipeline)``  and use the 2 catalogs. I am not sure I fully understand the implication, so this may not be a good idea, but this API seems easier to use for me if we want to enable to interact with the catalog as a standalone component. 

3. ``kedro catalog list`` -> ``def list_catalog_datasets(self, pipelines: list[str] | None = None) -> dict:`` 
This is basically "just a filter" (not really, but still) specific to the session, so I'd be in favor of lettting it at the session level because it is very session specific.

**Comment by datajoely** [comment]
> This is painful to use interactively because the session encapsulates a lot of thing and you lose finer control, and I've seen a lot of user use pipeline + catalog + runner in interactive mode rather than using the session directly. There are also a lot of plugins using dataset.exist() to materialize dataset factories in a catalog instance, and I am not sure they will be able to do it easily without creating a session. I think it would be much easier to interact with the standalone catalog directly.

Couldn't agree with this more, the last project I was on at QB (MA) completely bypassed it. 

> Using the session is a pain for testing because you need a lot of monkeypatching, and using catalog directly is much easier.

On the testing point I would also ask that Kedro actually provide a testing framework like other [large libraries provide this sort of thing](https://pandas.pydata.org/docs/reference/testing.html).

**Comment by ElenaKhaustova** [comment]
@datajoely, @Galileo-Galilei, thank you both for sharing the feedback. It is so valuable and made me think that we should find a way to make these commands part of the catalog API.

I still want to avoid coupling catalog and pipelines and I came to the solution where we create catalog mixins instead of session mixins. We can keep them separately from the catalog implementation and inject them dynamically at the time catalog is created, like that:

```python
class Catalog:
    """Our usual catalog without tightness to pipelines"""
    def __init__(self, a: int):
        self.a = a

    def print_attr(self):
        print(self.a)


class CatalogMixin:
    """Any extension to the catalog API that we want"""
    a: int

    def print_sum(self):
        print(self.a + self.a)

    def print_mult(self):
        print(self.a * self.a)


def compose_classes(base, *mixins):
    """Injecting mixins dynamically"""
    t = (base, ) + mixins
    name = f"{base.__name__}With{''.join(x.__name__ for x in mixins)}"
    return type(name, t, {})


def main():
    # At the time we create a catalog in the session we can inject all the mixins
    # Mixins will live separately from the catalog implementation and outside io module
    # So KedroCatalog is not tight with pipelines
    # In the pipeline + catalog + runner setup users will decide on their own
    # whether they instantiate catalog with mixins or not
    catalog = compose_classes(Catalog, CatalogMixin)(a=3)
    catalog.print_attr()
    catalog.print_sum()
    catalog.print_mult()


if __name__ == "__main__":
    main()
```

```console
>> python test_mixins.py
3
6
9
```

WDYT?

**Comment by ElenaKhaustova** [comment]
## That's how it looks in the interactive mode

### List patterns (formal kedro catalog rank)
```console
kedro ipython

In [1]: catalog.list_patterns()
Out[1]: ['{name}.{folder}#csv', '{name}_data', 'out-{dataset_name}', '{dataset_name}#csv', 'in-{dataset_name}']
```

### Resolve patterns (formal kedro catalog list)

```console
kedro ipython

In [1]: catalog.resolve_patterns()
Out[1]:
{
    'X_test': {'type': 'kedro.io.memory_dataset.MemoryDataset', 'copy_mode': None},
    'companies#csv': {
            'type': 'pandas.CSVDataset',
            'filepath': '/Projects/kedro-tests/default/data/01_raw/companies.csv'
        },
    'shuttle_id_dataset': {
            'type': 'kedro_datasets.pandas.sql_dataset.SQLQueryDataset',
            'sql': 'select shuttle, shuttle_id from spaceflights.shuttles;',
            'credentials': 'shuttle_id_dataset_credentials',
            'execution_options': {'stream_results': True},
            'load_args': {'chunksize': 1000},
            'fs_args': None,
            'filepath': None
        },
}

In [2]: catalog["new_dataset"] = 123

In [3]: catalog.resolve_patterns()
Out[3]:
{
    'X_test': {'type': 'kedro.io.memory_dataset.MemoryDataset', 'copy_mode': None},
    'companies#csv': {
            'type': 'pandas.CSVDataset',
            'filepath': '/Projects/kedro-tests/default/data/01_raw/companies.csv'
        },
    'shuttle_id_dataset': {
            'type': 'kedro_datasets.pandas.sql_dataset.SQLQueryDataset',
            'sql': 'select shuttle, shuttle_id from spaceflights.shuttles;',
            'credentials': 'shuttle_id_dataset_credentials',
            'execution_options': {'stream_results': True},
            'load_args': {'chunksize': 1000},
            'fs_args': None,
            'filepath': None
        },
    'new_dataset': {'type': 'kedro.io.memory_dataset.MemoryDataset', 'copy_mode': None},
}
```

### List datasets (formal kedro catalog resolve)

```console
kedro ipython

In [1]: catalog.list_datasets(pipelines=["data_processing"])
Out[1]:

{
    'data_processing': {
        'datasets': {
            'kedro_datasets.pandas.parquet_dataset.ParquetDataset': ['model_input_table', 'preprocessed_shuttles'],
            'kedro_datasets.pandas.excel_dataset.ExcelDataset': ['shuttles']
        },
        'factories': {'kedro_datasets.pandas.csv_dataset.CSVDataset': ['reviews.01_raw#csv', 'companies#csv']},
        'defaults': {'kedro.io.memory_dataset.MemoryDataset': ['preprocessed_companies']}
    }
}

In [2]: catalog["companies#csv"]
Out[2]: kedro_datasets.pandas.csv_dataset.CSVDataset(filepath=PurePosixPath('/Projects/kedro-tests/default/data/01_raw/companies.csv'), protocol='file', load_args={}, save_args={'index': False})

In [3]: scatalog.list_datasets(pipelines=["data_processing"])])
Out[3]:

{
    'data_processing': {
        'datasets': {
            'kedro_datasets.pandas.parquet_dataset.ParquetDataset': ['model_input_table', 'preprocessed_shuttles'],
            'kedro_datasets.pandas.excel_dataset.ExcelDataset': ['shuttles'],
            'kedro_datasets.pandas.csv_dataset.CSVDataset': ['companies#csv']
        },
        'factories': {'kedro_datasets.pandas.csv_dataset.CSVDataset': ['reviews.01_raw#csv']},
        'defaults': {'kedro.io.memory_dataset.MemoryDataset': ['preprocessed_companies']}
    }
}
```

### pipeline + catalog + runner workflow

```python
import yaml
from kedro.framework.context import CatalogCommandsMixin, compose_classes
from kedro.io import KedroDataCatalog
from kedro.pipeline import node
from kedro.pipeline.modular_pipeline import pipeline
from kedro.runner import SequentialRunner


def identity(arg):
    return arg


def create_pipeline():
    dummy_pipeline = pipeline(
        [
            node(
                func=identity,
                inputs="node_1_input",
                outputs="node_1_output",
                name="node_1",
            ),
            node(
                func=identity,
                inputs="node_1_output",
                outputs="node_2_output",
                name="node_2",
            ),
        ]
    )
    return dummy_pipeline


def create_data_catalog():
    catalog_conf = """
'node_1_input':
    type: pandas.CSVDataset
    filepath: data/01_raw/companies.csv

'node_2_{pattern}':
    type: pandas.CSVDataset
    filepath: data/test_output/'node_2_{pattern}'.csv
"""
    # Use compose_classes
    catalog_class = compose_classes(KedroDataCatalog, CatalogCommandsMixin)

    # Or create catalog with mixins
    # class KedroDataCatalogWithMixins(KedroDataCatalog, CatalogCommandsMixin):
    #     pass

    catalog_conf = yaml.safe_load(catalog_conf)
    catalog = catalog_class.from_config(catalog_conf)
    return catalog


def main():
    catalog = create_data_catalog()
    test_pipeline = create_pipeline()

    print(catalog.list_patterns())
    print(catalog.list_datasets(pipelines=[test_pipeline]))
    print(catalog.resolve_patterns(pipelines=[test_pipeline]))

    runner_obj = SequentialRunner()
    runner_obj.run(test_pipeline, catalog=catalog)


if __name__ == "__main__":
    main()
```

```console
['node_2_{pattern}']

{'pipeline_0': {
    'datasets': {'kedro_datasets.pandas.csv_dataset.CSVDataset': ['node_1_input']}, '
    factories': {'kedro_datasets.pandas.csv_dataset.CSVDataset': ['node_2_output']}, 
    'defaults': {'kedro.io.memory_dataset.MemoryDataset': ['node_1_output']}}
}

{'node_1_input': {
    'type': 'kedro_datasets.pandas.csv_dataset.CSVDataset', 
    'filepath': 'data/01_raw/companies.csv', 
    'load_args': None, 
    'save_args': None,
    'credentials': None, 
    'fs_args': None
}, 
'node_2_output': {'type': 'pandas.CSVDataset', 'filepath': "data/test_output/'node_2_output'.csv"}}
```

**Comment by datajoely** [comment]
I really like this, is there going to be a similar thing for `catalog.list_parameters`? i.e. can we see the defaults before interpolation.

**Comment by ElenaKhaustova** [comment]
> I really like this, is there going to be a similar thing for `catalog.list_parameters`? i.e. can we see the defaults before interpolation.

yeah, it's a good idea to add it, thank you!

**Comment by ravi-kumar-pilla** [comment]
Hi @ElenaKhaustova , 

The `Mixin` approach looks great. After going through the discussions and suggestions, I have few questions based on my understanding -

1. `catalog.list_patterns()` - Does this also have a `pipelines` arg ?

2. `catalog.list_datasets()` - Can we accept another arg here like `resolve_patterns` which takes in a list of specified patterns to resolve and may be a `resolve_all_patterns` arg which if `True` eagerly resolves patterns irrespective of the dataset is accessed or not ? From user perspective wouldn't this be an easy way to list `datasets` (i.e., both factory and regular) instead of doing another step of `catalog["companies#csv"]` i.e., accessing the dataset pattern to get it under the datasets list ?

3. The json structure for `catalog.resolve_patterns` has `dataset_name` as key which is intuitive if user wants to iterate over. The json structure for `catalog.list_datasets` is a bit complex (atleast on the first glance, considering a naive user). Wouldn't it be nice to keep the json structure similar to catalog config yaml structure we have ? This will be intuitive and yes it would make the json large (may be introduce a count arg for the list_datasets() ? ). Wouldn't this be a common use case if the user is iterating over the `catalog.list_datasets` json ?

As a dev in general if I see a list or dict, my most common next step is to iterate and analyze each item. Based on that, I felt (3) a bit involved. On the contrary, I understand that grouping clearly reduces the size and may be easy to navigate. 

Thank you

**Comment by Galileo-Galilei** [comment]
I've looked at the implementation, and I am actually a bit confused about the need to have a defaut behaviour  resolve_patterns if no pipeline is provided. When ``None`` is passed to the pipelines arguments (``catalog.resolve_patterns(pipelines=None)``), you import the pipelines under the hood (``from kedro.framework.project import pipelines as _pipelines``) to ensure the function works by default. 

If I understand the code correctly, this is the *only* things that couples catalog and pipelines implementation, and it should be fairly easy to decouple by forcing the user to provide the pipelines themselves. I understand all the mixin implementation which tries to decouples this import from the base class, but the real question is : should it really behaves likes this by default? I think anyone who is trying to use this function directly is supposed to have access (exactly like what you are doing here i.e. by importing the pipeline) to the pipeline to be resolved against so I don't really see the point to do it automatically at the cost of coupling components. 

There may have implications I don't fully grasp, e.g. :
- is there somewhere in the standard ``kedro run``  workflow where we try to access the resolved catalog without having the pipeline accessible ? 
- are there some tricky implications when resolving between thread/ cores if using the ``ThreadRunner`` or ``ParallelRunner`` (but even if the answer is yes, how your implementation is solving this?)

What do other thinks about it? 

Otherwise, everything looks great!

**Comment by ElenaKhaustova** [comment]
@ravi-kumar-pilla thank you for following this long read and leaving the feedback! Here are some answers to what you asked:

1. Yes
2. We want to keep `datasets`, `factories` and `defaults` separately, so it is what's the difference between them. They are resolved in the output representation but the catalog is not modified with these resolutions.
3. That's a valid point and I think we should consider this in the final implementation but my initial thought was to clarify the difference between `datasets`, `factories` and `defaults`.

**Comment by ElenaKhaustova** [comment]
@Galileo-Galilei good points

Technically, we could accept just pipeline objects as outputs to these methods so that users load them programmatically or, in the interactive environment, we load them using `_pipelines `. But I think providing pipeline objects to the catalog methods is already coupling because we will expect the specific type and the implementation, so we'll have to import `Pipeline` type and use pipeline API inside catalog.  So it will break this contract `CLI > Context > Library, Runner > Extras > IO & Pipeline` and it won't differ from importing `_pipelines ` directly.

Otherwise, there are no other implications.

**Comment by astrojuanlu** [comment]
Similarly, what PR closed this?

**Comment by ElenaKhaustova** [comment]
> Similarly, what PR closed this?

https://github.com/kedro-org/kedro/pull/4590
